File: trunk/managed/Warp3D/Warp3DVB/warp_Vertex.vb
' Created by: X
' http://www.createdbyx.com/
' ----------------------
' C# to VB.NET conversion of this code was done by hand and with the assistance of the 
' Convert C# to VB .NET application availible at http://www.kamalpatel.net/ConvertCSharp2VB.aspx
' ConvertCSharp2VB utility was developed by: Kamal Patel (http://www.KamalPatel.net) 
' ----------------------
' Thanks to Alan Simes ( http://alansimes.blogdns.net/ ) for creating the Warp3D library.
' http://alansimes.blogdns.net/cs/files/default.aspx
' ----------------------

Imports System
Imports System.Collections

namespace Rednettle.Warp3D
    Public Class warp_Vertex
        Public parent As warp_Object

        Public pos As New warp_Vector() '   //(x,y,z) Coordinate of vertex
        Public pos2 As warp_Vector '  //Transformed vertex coordinate
        Public n As New warp_Vector() '   //Normal Vector at vertex
        Public n2 As warp_Vector '  //Transformed normal vector (camera space)

        Public x As Integer '  //Projected x coordinate
        Public y As Integer '  //Projected y coordinate
        Public z As Integer '  //Projected z coordinate for z-Buffer

        Public u As Single = 0 ' // Texture x-coordinate (relative)
        Public v As Single = 0 ' // Texture y-coordinate (relative)

        Public nx As Integer = 0 ' // Normal x-coordinate for envmapping
        Public ny As Integer = 0 ' // Normal y-coordinate for envmapping
        Public tx As Integer = 0 ' // Texture x-coordinate (absolute)
        Public ty As Integer = 0 ' // Texture y-coordinate (absolute)

        Public sw As Single = 1.0F


        Public visible As Boolean = True '  //visibility tag for clipping
        Public clipcode As Integer = 0
        Public id As Integer ' // Vertex index

        Private fact As Single
        'private Vector neighbor=new Vector(); //Neighbor triangles of vertex
        Private neighbor As New ArrayList()



        Public Sub New()
            pos = New warp_Vector(0.0F, 0.0F, 0.0F)
        End Sub

        Public Sub New(ByVal xpos As Single, ByVal ypos As Single, ByVal zpos As Single)
            pos = New warp_Vector(xpos, ypos, zpos)
        End Sub

        Public Sub New(ByVal xpos As Single, ByVal ypos As Single, ByVal zpos As Single, ByVal u As Single, ByVal v As Single)
            pos = New warp_Vector(xpos, ypos, zpos)
            Me.u = u
            Me.v = v
        End Sub

        Public Sub New(ByVal ppos As warp_Vector)
            pos = ppos.getClone()
        End Sub

        Public Sub New(ByVal ppos As warp_Vector, ByVal u As Single, ByVal v As Single)
            pos = ppos.getClone()
            Me.u = u
            Me.v = v
        End Sub


        Public Sub project(ByVal vertexProjection As warp_Matrix, ByVal normalProjection As warp_Matrix, ByVal camera As warp_Camera)
            ' Projects this vertex into camera space
            pos2 = pos.transform(vertexProjection)
            n2 = n.transform(normalProjection)

            fact = camera.screenscale / camera.fovfact / CSng(IIf(pos2.z > 0.1, pos2.z, 0.1F))
            x = CInt((pos2.x * fact + (camera.screenwidth >> 1)))
            y = CInt((-pos2.y * fact + (camera.screenheight >> 1)))
            z = CInt((65536.0F * pos2.z))
            sw = -(pos2.z)
            nx = CInt((n2.x * 127 + 127))
            ny = CInt((n2.y * 127 + 127))
            If parent.material Is Nothing Then Return
            If parent.material.texture Is Nothing Then Return
            tx = CInt((parent.material.texture.width * u))
            ty = CInt((parent.material.texture.height * v))
        End Sub

        Public Sub setUV(ByVal u As Single, ByVal v As Single)
            Me.u = u
            Me.v = v
        End Sub


        Public Sub clipFrustrum(ByVal w As Integer, ByVal h As Integer)
            ' View plane clipping
            clipcode = 0
            If x < 0 Then
                clipcode = clipcode Or 1
            End If
            If x >= w Then
                clipcode = clipcode Or 2
            End If
            If y < 0 Then
                clipcode = clipcode Or 4
            End If
            If y >= h Then
                clipcode = clipcode Or 8
            End If
            If pos2.z < 0 Then
                clipcode = clipcode Or 16
            End If
            visible = (clipcode = 0)
        End Sub

        Public Sub registerNeighbor(ByVal triangle As warp_Triangle)
            If Not neighbor.Contains(triangle) Then
                neighbor.Add(triangle)
            End If
        End Sub

        Public Sub resetNeighbors()
            neighbor.Clear()
        End Sub

        Public Sub regenerateNormal()
            Dim nx As Single = 0
            Dim ny As Single = 0
            Dim nz As Single = 0
            Dim Enumerator As IEnumerator = neighbor.GetEnumerator()
           
            Dim tri As warp_Triangle
            Dim wn As warp_Vector
            While Enumerator.MoveNext()
                tri = CType(Enumerator.Current, warp_Triangle)
                wn = tri.getWeightedNormal()
                nx += wn.x
                ny += wn.y
                nz += wn.z
            End While

            n = New warp_Vector(nx, ny, nz).normalize()
        End Sub

        '/*
        '		public void regenerateNormal()
        '		{
        '			float nx=0;
        '			float ny=0;
        '			float nz=0;
        '			Enumeration enum=neighbor.elements();
        '			warp_Triangle tri;
        '			warp_Vector wn;
        '			while (enum.hasMoreElements())
        '			{	
        '				tri=(warp_Triangle)enum.nextElement();
        '				wn=tri.getWeightedNormal();
        '				nx+=wn.x;
        '				ny+=wn.y;
        '				nz+=wn.z;
        '			}

        '			n=new warp_Vector(nx,ny,nz).normalize();
        '		}
        '*/

        Public Sub scaleTextureCoordinates(ByVal fx As Single, ByVal fy As Single)
            u *= fx
            v *= fy
        End Sub

        Public Sub moveTextureCoordinates(ByVal fx As Single, ByVal fy As Single)
            u += fx
            v += fy
        End Sub

        Public Function getClone() As warp_Vertex
            Dim NewVertex As warp_Vertex = New warp_Vertex()
            NewVertex.pos = pos.getClone()
            NewVertex.n = n.getClone()
            NewVertex.u = u
            NewVertex.v = v

            Return NewVertex
        End Function


        Public Function equals(ByVal v As warp_Vertex) As Boolean
            Return ((pos.x = v.pos.x) And (pos.y = v.pos.y) And (pos.z = v.pos.z))
        End Function

        Public Function equals(ByVal v As warp_Vertex, ByVal tolerance As Single) As Boolean
            Return Math.Abs(warp_Vector.sub(pos, v.pos).length()) < tolerance
        End Function
    End Class
End Namespace

File: trunk/managed/Warp3Dmod/warp_Vertex.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace Warp3D
{
    /// <summary>
    /// Summary description for warp_Vertex.
    /// </summary>
    public class warp_Vertex
    {
        public warp_Object parent;

        public warp_Vector pos = new warp_Vector();   //(x,y,z) Coordinate of vertex
        public warp_Vector n = new warp_Vector();   //Normal Vector at vertex
        public float u = 0; // Texture x-coordinate (relative)
        public float v = 0; // Texture y-coordinate (relative)


        public warp_Vector pos2;  //Transformed vertex coordinate
        public warp_Vector n2;  //Transformed normal vector (camera space)
        public int x;  //Projected x coordinate
        public int y;  //Projected y coordinate
        public int z;  //Projected z coordinate for z-Buffer

        public float tx = 0; // Texture x-coordinate (relative)
        public float ty = 0; // Texture y-coordinate (relative)

        public int nx = 0; // Normal x-coordinate for envmapping
        public int ny = 0; // Normal y-coordinate for envmapping

        public float invZ = 1.0f;

        public bool visible = true;  //visibility tag for clipping
        public int clipcode = 0;
        public int id; // Vertex index

        //private Vector neighbor=new Vector(); //Neighbor triangles of vertex
        private List<warp_Triangle> neighbor = new List<warp_Triangle>();

        public warp_Vertex()
        {
            pos = new warp_Vector(0f, 0f, 0f);
        }

        public warp_Vertex(float xpos, float ypos, float zpos)
        {
            pos = new warp_Vector(xpos, ypos, zpos);
        }

        public warp_Vertex(float xpos, float ypos, float zpos, float u, float v)
        {
            pos = new warp_Vector(xpos, ypos, zpos);
            this.u = u;
            this.v = v;
        }

        public warp_Vertex(warp_Vector ppos)
        {
            pos = ppos.getClone();
        }

        public warp_Vertex(warp_Vector ppos, float u, float v)
        {
            pos = ppos.getClone();
            this.u = u;
            this.v = v;
        }

        public warp_Vertex(warp_Vector ppos, warp_Vector norm, float u, float v)
        {
            pos = ppos.getClone();
            n = norm.getClone();
            this.u = u;
            this.v = v;
        }

        public void project(warp_Matrix vertexProjection, warp_Matrix normalProjection, warp_Camera camera)
        // Projects this vertex into camera space
        {
            pos2 = pos.transform(vertexProjection);
            n2 = n.transform(normalProjection);

            if(pos2.z < 0.001f && pos2.z > -0.0001f)
                pos2.z = 0.001f;

            if(camera.isOrthographic)
            {
                x = (int)pos2.x;
                y = (int)pos2.y;
                invZ = -1.0f;
                tx = -u;
                ty = -v;
            }
            else
            {
                invZ = 1.0f / pos2.z;
                x = (int)(pos2.x * invZ + camera.halfscreenwidth);
                y = (int)(pos2.y * invZ + camera.halfscreenheight);
                invZ = - invZ;
                tx = u * invZ;
                ty = v * invZ;
            }

            z = (int)(65536f * pos2.z);
            nx = ((int)(n2.x * 127 + 127)) << 16;
            ny = ((int)(n2.y * 127 + 127)) << 16;
        }

        public void setUV(float u, float v)
        {
            this.u = u;
            this.v = v;
        }

        public void clipFrustrum(int w, int h)
        {
            // View plane clipping
            clipcode = 0;
            if(x < 0)
                clipcode |= 1;
            else if(x >= w)
                clipcode |= 2;
            if(y < 0)
                clipcode |= 4;
            else if(y >= h)
                clipcode |= 8;
            if(pos2.z < 0)
                clipcode |= 16;
            visible = (clipcode == 0);
        }

        public void registerNeighbor(warp_Triangle triangle)
        {
            if(!neighbor.Contains(triangle))
                neighbor.Add(triangle);
        }

        public void resetNeighbors()
        {
            neighbor.Clear();
        }

        public void regenerateNormal()
        {
            float nx = 0;
            float ny = 0;
            float nz = 0;
            IEnumerator enumerator = neighbor.GetEnumerator();

            warp_Triangle tri;
            warp_Vector wn;
            while(enumerator.MoveNext())
            {
                tri = (warp_Triangle)enumerator.Current;
                wn = tri.getWeightedNormal();
                nx += wn.x;
                ny += wn.y;
                nz += wn.z;
            }

            n = new warp_Vector(nx, ny, nz).normalize();
        }

        /*
                public void regenerateNormal()
                {
                    float nx=0;
                    float ny=0;
                    float nz=0;
                    Enumeration enum=neighbor.elements();
                    warp_Triangle tri;
                    warp_Vector wn;
                    while (enum.hasMoreElements())
                    {	
                        tri=(warp_Triangle)enum.nextElement();
                        wn=tri.getWeightedNormal();
                        nx+=wn.x;
                        ny+=wn.y;
                        nz+=wn.z;
                    }

                    n=new warp_Vector(nx,ny,nz).normalize();
                }
        */
        public void scaleTextureCoordinates(float fx, float fy)
        {
            u *= fx;
            v *= fy;
        }

        public void moveTextureCoordinates(float fx, float fy)
        {
            u += fx;
            v += fy;
        }

        public warp_Vertex getClone()
        {
            warp_Vertex newVertex = new warp_Vertex();
            newVertex.pos = pos.getClone();
            newVertex.n = n.getClone();
            newVertex.u = u;
            newVertex.v = v;

            return newVertex;
        }

        public bool equals(warp_Vertex v)
        {
            return ((pos.x == v.pos.x) && (pos.y == v.pos.y) && (pos.z == v.pos.z));
        }

        public bool equals(warp_Vertex v, float tolerance)
        {
            return Math.Abs(warp_Vector.sub(pos, v.pos).length()) < tolerance;
        }
    }
}

